operator-framework/src/main/java/io/javaoperatorsdk/operator/processing/event/internal/CustomResourceEventSource.java

```diff
@@ -0,0 +1,92 @@
+
+/**
+ * This is a special case since is not bound to a single custom resource
+ */
+public class CustomResourceEventSource extends AbstractEventSource implements Watcher<CustomResource> {
+
+    private final static Logger log = LoggerFactory.getLogger(CustomResourceEventSource.class);
+
+    private final ResourceCache resourceCache;
+    private MixedOperation client;
+    private final String[] targetNamespaces;
+
+    public static CustomResourceEventSource customResourceEventSourceForAllNamespaces(ResourceCache resourceCache,
+                                                                                      MixedOperation client) {
+        return new CustomResourceEventSource(resourceCache, client, null);
+    }
+
+    public static CustomResourceEventSource customResourceEventSourceForTargetNamespaces(ResourceCache resourceCache,
+                                                                                         MixedOperation client,
+                                                                                         String[] namespaces) {
+        return new CustomResourceEventSource(resourceCache, client, namespaces);
+    }
+
+    private CustomResourceEventSource(ResourceCache resourceCache, MixedOperation client, String[] targetNamespaces) {
+        this.resourceCache = resourceCache;
+        this.client = client;
+        this.targetNamespaces = targetNamespaces;
+    }
+
+    private boolean isWatchAllNamespaces() {
+        return targetNamespaces == null;
+    }
+
+    public void addedToEventManager() {
+        registerWatch();
+    }
+
+    private void registerWatch() {
+        CustomResourceOperationsImpl crClient = (CustomResourceOperationsImpl) client;
+        if (isWatchAllNamespaces()) {
+            crClient.inAnyNamespace().watch(this);
+        } else if (targetNamespaces.length == 0) {
+            client.watch(this);
+        } else {
+            for (String targetNamespace : targetNamespaces) {
+                crClient.inNamespace(targetNamespace).watch(this);
+                log.debug("Registered controller for namespace: {}", targetNamespace);
+            }
+        }
+    }
+
+    @Override
+    public void eventReceived(Watcher.Action action, CustomResource customResource) {
+        log.debug("Event received for action: {}, resource: {}",
+                customResource.getMetadata().getName(), customResource);
+        resourceCache.cacheResource(customResource); // always store the latest event. Outside the sync block is intentional.
+        if (action == Action.ERROR) {
+            log.debug("Skipping {} event for custom resource uid: {}, version: {}", action,
+                    getUID(customResource), getVersion(customResource));
+            return;
+        }
+        eventHandler.handleEvent(new CustomResourceEvent(action, customResource, this));
+    }
+
+    @Override
+    public void onClose(KubernetesClientException e) {
+        if (e.getCode() == HTTP_GONE) {
```

Comment:
Based on the Watcher.onclose javadoc, on normal closes the exception is null.
```suggestion
        if (e == null) {
            return;
        }
        if (e.getCode() == HTTP_GONE) {
```